text/0314-cdk-construct-hub.md

```diff
@@ -0,0 +1,310 @@
+to prefetch all information necessary to back the webiste's search and package
+detail pages.
+
+#### Back-End
+
+1. The back-end for Construct Hub starts with a custom-built event source,
+   which implements the logic necessary to detect new packages that are relevant
+   to the Construct Hub, and sends messages to an SQS queue for further
+   processing. Notifications have the following attributes:
+
+   Name        | Description
+   ------------|----------------------------------------------------------------
+   `assembly`  | The `.jsii` assembly included in the package
+   `time`      | The timestamp at which the version was created
+   `integrity` | An integirity check for the complete record
+
+   In cases where the `assembly` object is too large, it may be staged in S3
+   instead of being sent directly as part of the message payload. In such cases,
+   the `assembly` field contains an S3 URI instead of a JSON object.
+
+1. A Lambda function then picks up messages from the SQS queue and prepares the
+   artifacts consumed by the front-end application, stored in a dedicated S3
+   bucket using the following key format:
+   `assemblies/${assembly.name}/v${assembly.version}/assembly.json`
+
+   This function validates the contents of the `assembly` to ensure the message
+   was well-formed. In case the message is found to be incoherent, it is sent to
+   a dead-letter queue and an alarm is triggered.
+
+   It also creates or updates the `assemblies/${assembly.name}/versions.json`
+   object, which contains a list of all versions ever indexed for the package,
+   matched to the versions' current status (`deprecated`, `latest`, ...):
+
+   Name        | Description
+   ------------|----------------------------------------------------------------
+   `v${major}` | All versions part of this major version stream, and their status
+
+   [SemVer]: https://semver.org/spec/v2.0.0.html
+
+1. A series of Lambda functions prepare language-specific assembly files for
+   each configured language, with adjusted naming conventions, and updated
+   sample code fragments, and proceeds to store those at:
+   `assemblies/${assembly.name}/v${assembly.version}/assembly-${lang}.json`.
+   This transformation is backed by the `jsii-rosetta` tool, which is part of
+   the [jsii project][jsii].
+
+   [jsii]: https://aws.github.io/jsii
+
+1. A Lambda function keeps the `latest.json` object updated with the last 20
+   package versions indexed (according to the `time` field reported by the
+   ingestion component).
+
+1. A Lambda function keeps the `catalog.json` object updated with the latest
+   versions of each package's major version lines, which backs the websites'
+   search page. The object may be sharded if it becomes too large.
+
+#### Front-End
+
+The front-end consists in a simple [React] application hosted in an S3 bucket.
+A CloudFront distribution fronts this application, ensuring it is served from
```

Comment:
Nextjs, Tailwind, TS or JS ?

Reply:
I'm getting into *some* aspects of technical decisions, but trying to avoid closing too many doors this early. I am here to describe what will be done, not how it will be done (at least - implementors get to have some freedom of movement).